5.auth-service/index.js

```diff
@@ -1,10 +1,351 @@
+    resData.serverError = "serverError";
+    res.send(resData);
+    return;
+  };
+
+  if(result.length > 0){
+    let userResult = result[0];      
+    const user = {userName: userResult.user_name,  userId: userResult.user_id, accountId: userResult.account_id};
+    resData.user_name = userResult.user_name;
+    resData.accessToken = sessionHelper.createSession(user);
+
+  } else {
+    resData.valid = false;
+    resData.serverError = 'IncorrectMailOrPassword';
+    console.log('here in incorrect mail');
+  }
+  res.send(resData);
+});
+
+/**
+ * Checks the user is logged in and deletes his session.
+ */
+app.post('/logout', function (req, res) {
+  let resData = { valid: false }
+
+  var userData = sessionHelper.verifySession(req.headers.authorization);
+  if(userData){
+    sessionHelper.deleteSession(userData);
+    resData.valid = true;
+  }
+  res.send(resData);
+});
+
+/**
+ * @param {string} password 
+ * @returns the encoded password
+ */
+function encodePassword(password){
+  return md5(password);
+}
+
+/**
+ * @param {request object} req 
+ * @param {response object} res 
+ * @param {function} next 
+ * @returns false if the token not verified
+ */
+function authMiddleware(req, res, next){
+  const userData = sessionHelper.verifySession(req.headers.authorization);
+  console.log(userData)
+  if(userData){
+    next()
+  } else {
+      console.log("returns false");
+      // TODO return something else
+      return res.send(false);
+  }
+}
+
+/**
+ * Sets the response to true.
+ */
+app.get('/ping', authMiddleware,  function(req, res){
+  res.send(true);
+});
+
+/**
+ * Sends an email to the given mail address with a link to reset the user's password.
+ */
+app.post('/forgotPassword', async function (req, res) {
+  
+  let resData = {valid: true, errors: [], serverError: ''};
+
+  // fields validations
+  resData =  validateAll(req.body);
+  if(!resData.valid){
+    res.send(resData);
+  }
+
+  const userMail = req.body.mail.value;
+  if(userMail){
+      // Check if the mail is in the db
+      var sql = `SELECT * FROM users WHERE user_mail = '${userMail}'`;
+      let result = await sqlHelper.select(sql).catch((err)=>{});
+
+      // The query failed
+      if(!result){
+        resData.valid = false;
+        resData.serverError = "serverError";
+        res.send(resData);
+        return;
+      };
+
+      if(result.length == 0){
+        resData.valid = true;
+        res.send(resData);
+        return;
+      }
+
+      // User exist
+      resData.valid = true;
+      // Encoding the mail address
+      let tokenBody = {userMail: userMail};
+      const mailToken = sessionHelper.createToken(tokenBody);
+      console.log("this is the mail link to reset the password: ", `http://localhost:3000/resetPassword/${mailToken}`);
+      try {
+        resData = await sendMail('coheen1@gmail.com', userMail, 'RGB - Reset password', `<a href=${`http://localhost:3000/resetPassword/${mailToken}`}>Click to reset your password.</a>`);
```

Comment:
take http://localhost:3000/ out to a constant (in you env file - because this is a changeable value)